File: pitchroll3/app/src/main/java/edu/cs4730/pitchroll3/MainActivity.java
package edu.cs4730.pitchroll3;

import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;

import java.util.Locale;

import edu.cs4730.pitchroll3.databinding.ActivityMainBinding;


/**
 * http://stackoverflow.com/questions/9454948/android-pitch-and-roll-issue
 * modified to remove the Orientation sensor that google dec in api 9.
 */

public class MainActivity extends AppCompatActivity implements SensorEventListener {

    private static final String TAG = "MainActivity";
    private SensorManager mgr;
    private Sensor accel;
    private Sensor compass;

    private ActivityMainBinding binding;
    private boolean ready = false;
    private float[] accelValues = new float[3];
    private float[] compassValues = new float[3];
    private float[] inR = new float[9];
    private float[] inclineMatrix = new float[9];
    private float[] prefValues = new float[3];
    private float mAzimuth;
    private double mPitch;
    private double mInclination;
    private int counter;
    private int mRotation;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        mgr = (SensorManager) this.getSystemService(SENSOR_SERVICE);
        accel = mgr.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        compass = mgr.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD);

        mRotation = rotationInfo();
    }

    public int rotationInfo() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) {
            return getDisplay().getRotation();
        } else {

            return getWindowManager().getDefaultDisplay().getRotation();
        }
    }

    @Override
    protected void onResume() {
        mgr.registerListener(this, accel, SensorManager.SENSOR_DELAY_NORMAL);
        mgr.registerListener(this, compass, SensorManager.SENSOR_DELAY_NORMAL);
        super.onResume();
    }

    @Override
    protected void onPause() {
        mgr.unregisterListener(this, accel);
        mgr.unregisterListener(this, compass);
        super.onPause();
    }

    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        // ignore
    }

    public void onSensorChanged(SensorEvent event) {
        // Need to get both accelerometer and compass
        // before we can determine our orientationValues
        // Log.wtf(TAG, "onSensorChanged.");
        switch (event.sensor.getType()) {
            case Sensor.TYPE_ACCELEROMETER:
                for (int i = 0; i < 3; i++) {
                    accelValues[i] = event.values[i];
                }
                if (compassValues[0] != 0)
                    ready = true;
                break;
            case Sensor.TYPE_MAGNETIC_FIELD:
                for (int i = 0; i < 3; i++) {
                    compassValues[i] = event.values[i];
                }
                if (accelValues[2] != 0)
                    ready = true;
                break;
        }

        if (!ready) {
            binding.preferred.setText("Not enough data yet.");
            return;
        }
        if (SensorManager.getRotationMatrix(
            inR, inclineMatrix, accelValues, compassValues)) {
            // got a good rotation matrix
            SensorManager.getOrientation(inR, prefValues);
            mInclination = SensorManager.getInclination(inclineMatrix);
            // Display every 10th value
            //if(counter++ % 10 == 0) {
            doUpdate();
//				counter = 1;
            //}

        }
    }

    public void doUpdate() {
        if (!ready) {
            binding.preferred.setText("Not enough data yet2.");
            return;
        }
        // Log.wtf(TAG, "doUpdate.");
        mAzimuth = (float) Math.toDegrees(prefValues[0]);
        if (mAzimuth < 0) {
            mAzimuth += 360.0f;
        }
        //mPitch = 180.0  + Math.toDegrees(prefValues[1]); //so it goes from 0 to 360, instead of -180 to 180
        mPitch = Math.toDegrees(prefValues[1]);
        String msg = String.format(Locale.getDefault(),
            "Preferred:\nazimuth (Z): %7.3f \npitch (X): %7.3f\nroll (Y): %7.3f",
            mAzimuth, //heading
            mPitch,
            Math.toDegrees(prefValues[2]));
        binding.preferred.setText(msg);
    }

}

File: pitchroll3_tk/app/src/main/java/edu/cs4730/pitchroll3_tk/MainActivity.kt
package edu.cs4730.pitchroll3_tk

import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import androidx.appcompat.app.AppCompatActivity
import edu.cs4730.pitchroll3_tk.databinding.ActivityMainBinding
import java.util.Locale

class MainActivity : AppCompatActivity(), SensorEventListener {

    private val TAG = "MainActivity"
    private lateinit var mgr: SensorManager
    private var accel: Sensor? = null
    private var compass: Sensor? = null

    private lateinit var binding: ActivityMainBinding
    private var ready = false
    private val accelValues = FloatArray(3)
    private val compassValues = FloatArray(3)
    private val inR = FloatArray(9)
    private val inclineMatrix = FloatArray(9)
    private val prefValues = FloatArray(3)
    private var mAzimuth = 0f
    private var mPitch = 0.0
    private var mInclination = 0.0
    private val counter = 0
    private var mRotation = 0


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)


        mgr = this.getSystemService(SENSOR_SERVICE) as SensorManager
        accel = mgr.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        compass = mgr.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)
        val window = this.getSystemService(WINDOW_SERVICE) as WindowManager
        mRotation = window.defaultDisplay.rotation
    }

    override fun onResume() {
        mgr.registerListener(this, accel, SensorManager.SENSOR_DELAY_NORMAL)
        mgr.registerListener(this, compass, SensorManager.SENSOR_DELAY_NORMAL)
        super.onResume()
    }

    override fun onPause() {
        mgr.unregisterListener(this, accel)
        mgr.unregisterListener(this, compass)
        super.onPause()
    }


    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        // ignore
    }

    override fun onSensorChanged(event: SensorEvent) {
        // Need to get both accelerometer and compass
        // before we can determine our orientationValues
        // Log.wtf(TAG, "onSensorChanged.");
        when (event.sensor.type) {
            Sensor.TYPE_ACCELEROMETER -> {
                var i = 0
                while (i < 3) {
                    accelValues[i] = event.values[i]
                    i++
                }
                if (compassValues[0] != 0F) ready = true
            }

            Sensor.TYPE_MAGNETIC_FIELD -> {
                var i = 0
                while (i < 3) {
                    compassValues[i] = event.values[i]
                    i++
                }
                if (accelValues[2] != 0F) ready = true
            }
        }
        if (!ready) {
            binding.preferred.text = "Not enough data yet."
            return
        }
        if (SensorManager.getRotationMatrix(
                inR, inclineMatrix, accelValues, compassValues
            )
        ) {
            // got a good rotation matrix
            SensorManager.getOrientation(inR, prefValues)
            mInclination = SensorManager.getInclination(inclineMatrix).toDouble()
            // Display every 10th value
            //if(counter++ % 10 == 0) {
            doUpdate()
            //				counter = 1;
            //}
        }
    }

    fun doUpdate() {
        if (!ready) {
            binding.preferred.text = "Not enough data yet2."
            return
        }
        // Log.wtf(TAG, "doUpdate.");
        mAzimuth = Math.toDegrees(prefValues[0].toDouble()).toFloat()
        if (mAzimuth < 0) {
            mAzimuth += 360.0f
        }
        //mPitch = 180.0  + Math.toDegrees(prefValues[1]); //so it goes from 0 to 360, instead of -180 to 180
        mPitch = Math.toDegrees(prefValues[1].toDouble())
        val msg = String.format(
            Locale.getDefault(),
            "Preferred:\nazimuth (Z): %7.3f \npitch (X): %7.3f\nroll (Y): %7.3f",
            mAzimuth,  //heading
            mPitch,
            Math.toDegrees(prefValues[2].toDouble())
        )
        binding.preferred.text = msg
    }


}